proptest/Cargo.toml

```diff
@@ -57,6 +57,7 @@ break-dead-code = []
 
 [dependencies]
 bitflags = "1.0.1"
+unarray = "0.1.3"
```

Comment:
i'm going to be a bit nitpicky over this so feel free to push back.

How do you feel about moving this crate in to the proptest organization? Given how lean our dependencies are, I have a mild concern on taking this dependency given we wouldn't have any organizational control over updating it in the event we want or need to. It would also give us a chance to formally review the code as an org and make it a bit more public that we've done our due diligence with the unsafe portions of the code.

embedding a couple comments on the `unarray` code from a quick look through:

- `build_array_result` leaks if an error is encountered, even if it doesn't panic. A written MaybeUninit doesn't have its destructor run without a call to `assume_init` but we just return from the function without ever assuming any of the elements initialized
  - https://github.com/cameron1024/unarray/blob/92123dcb3bde1e4efd313da6fc0df0968d97b7a5/src/build.rs#L53
- same thing goes for `build_array_option` that return `None`
- same thing goes for `map_result` that returns `Err`
- same thing goes for `map_option` that returns `None`
- same thing goes for `from_iter` when the iter sizes don't match
- we could beef test coverage around `build_*` methods to include functions that return `Err` and `None`
- i saw a couple spots where you had `#[cfg_attr(miri, ignore)]`, so i'm assuming you've run this through miri?
- other than that, the unsafe code looked fine

Reply:
Regarding ownership, I'd be totally happy to move this into the proptest org. A slight bit of context, I usually have `#![deny(unsafe_code)]` in everything I work on, and always put all unsafe code into its own crate to force me to create proper abstractions. These were just some utilities I found myself writing a lot, so grateful to get some more eyes on it.

Regarding the technical issues:
 - the leaking thing sounds pretty bad, I'll definitely fix that. Initially when I read your comment I misread and thought you were talking about panicking, which seems harder to avoid in the generic case. But we can definitely track how many are initialized and drop them in the error cases
 - increasing test coverage sounds like a good idea
 - ironically, the `#[cfg_attr(miri, ignore)]`s are all for proptests :laughing: There's actually an open issue here for running proptests with miri, and I've been meaning to have a look, because IMO they should work well together (assuming performance doesn't become an issue).

But yeah, I'll push a new branch with some changes to `unarray`, and if we want to bring it into the org we can

Reply:
i don't feel super strongly about the ownership part -- mostly posed it as a straw-man proposal to get some input. Mainly just want to make sure we add this in a way that still feels "open" for consumers of proptest. Given its a new/relatively unknown crate developed by one of the maintainers and the maintainers are new, I don't want it to come off as self-promotion or anything like that (not saying that's what you're doing, i appreciate you willing to share the code and provide a solution that doesn't require a bump in MSRV).

Reply:
if we've gotten a few eyes on the crate and don't want to deal with any more overhead we can just move forward as is. If ever needed for maintenance in the future we could always pull parts of it in to proptest without any invasive changes

Reply:
Makes sense. How does this sound as a plan:
 - I'll push a fix to `unarray` for the leak issue you mentioned as well as some better test coverage. The fix I have in mind will fix the egregious cases, but won't cover leaking on panic (last time I had a look at this, it seems like this is not really solvable in the general case anyways)
 - we depend on the fixed version without doing any sort of ownership change for now
 - in the future, if we feel like we want to, we can chat about whether it makes sense to move it into the org (or just certain functions, maybe reimplement, change MSRV, etc, there's quite a few possibilities)

Does that sound good?

Reply:
that sounds great! agreed we shouldn't worry about panics

Reply:
I've got a PR here: https://github.com/cameron1024/unarray/pull/1/files . Feel free to have a look over the implementation. If you're happy with it I'll publish it as 0.1.4 and bump the version that `proptest` depends on

Reply:
I've bumped the version, let me know if you have any issues with it. AFAICT, it should be pretty solid now. Tested with miri, and added some tests to catch regressions for the leak issue